### DISCUSSION TITLE
I can't get REST API Data Source to work

### DISCUSSION THREAD

**Discussion by jeevisonline** [Budibase/budibase#4101] [Help]
Hello all,

I can't seem to get REST API data sources working properly.  I have tried connecting into a PSA with private/public keys with no success, and I have tried the public API https://scryfall.com/docs/api/ And I have only gotten 1 query working.

I'll talk about Scryfall first, because its public.

I have tried multiple get requests, but have only been able to get a single card's worth of data using the url 
`https://api.scryfall.com/cards/random`

I have tried to use the bulk-data, but the json file that I found isn't importing, either by URL or by FILE.

`https://c2.scryfall.com/file/scryfall-bulk/default-cards/default-cards-20220118220326.json`

I am new to APIs in general, so I think I am just doing something wrong.

**Comment by Rory-Powell** [comment]
Hi @jeevisonline. I think the problem here is with the endpoint returning too much data, the `default-cards-20220118220326.json` response is coming in at about 32MB which is a huge json response (around 65,000 keys), firefox and chrome themselves struggle to render this efficiently. After about 30 seconds I got a response in my budibase rest query but that then brings the ui to a halt as it processes all those rows.


I would suggest staying away from all of the bulk apis as they are all quite large. I did find this swagger definition for skryfall which can be used to import the api collection into budibase directly
https://raw.githubusercontent.com/jdharmon/scryfallapi/master/swagger.yaml

To import you can paste this url into the rest data import modal
<img width="699" alt="Screenshot 2022-01-19 at 12 37 35" src="https://user-images.githubusercontent.com/8755148/150132197-6760ea16-fb4e-455a-ac09-bf8d31e142ca.png">

which will generate all the requests 
<img width="1371" alt="Screenshot 2022-01-19 at 12 37 22" src="https://user-images.githubusercontent.com/8755148/150132317-1d052b3a-2ecc-4fd8-b16b-dd5504ca1dcb.png">

Where I think you might be running into trouble is how budibase handles URL parameters currently, I pasted a few of the URLs from the scryfall documentation and the request didn't work because the parameters weren't auto configured in the `Params` section of the rest query. I've added an enhancement for this here 
https://github.com/Budibase/budibase/issues/4106

If you use the import method from above the `Params` should be automatically configured for you and linked with `Bindings` that can be provided to the query from within your application
e.g. for `https://api.scryfall.com/cards/named` the `fuzzy` or `exact` param is required so I can provided a default value to it's binding and see the request succeed

<img width="990" alt="Screenshot 2022-01-19 at 12 45 13" src="https://user-images.githubusercontent.com/8755148/150133275-a270417f-acbd-40d4-a03a-51f7c844ec34.png">


Details on query import, params, bindings and more can be found here
https://docs.budibase.com/building-apps/data/tables/rest
